src/lib/fetch.ts

```diff
@@ -730,44 +730,61 @@ async function fetchActions(
       },
     };
   }
-
-  let actionMap = new Map<
-    string,
-    {
-      hash: string;
-      actions: string[][];
-    }
-  >();
-  for (const { actionState, actionData } of fetchedActions) {
-    for (const { accountUpdateId, data } of actionData) {
-      const action = actionMap.get(accountUpdateId);
-      if (action) {
-        const { actions } = action;
-        actions.push(data);
-        actionMap.set(accountUpdateId, {
-          hash: Ledger.fieldToBase58(Field(actionState)),
-          actions,
+  // Archive Node API returns actions in the latest order, so we reverse the array to get the actions in chronological order.
+  fetchedActions.reverse();
+  let actionsList: { actions: string[][]; hash: string }[] = [];
+  let latestActionsHash = SequenceEvents.emptySequenceState();
+
+  fetchedActions.forEach((fetchedAction) => {
+    const { actionState, actionData } = fetchedAction;
+    if (actionData.length === 0)
+      throw new Error(
+        `No action data was found for the account ${publicKey} with the latest action state ${actionState}`
+      );
+    let currentAccountUpdateId = actionData[0].accountUpdateId;
+    let currentActionList: string[][] = [];
+    actionData.forEach((action, i) => {
+      const { accountUpdateId, data } = action;
+      const isLastAction = i === actionData.length - 1;
+
+      if (accountUpdateId !== currentAccountUpdateId || isLastAction) {
+        if (isLastAction) currentActionList.push(data);
+        let actionsHash = SequenceEvents.hash(
+          currentActionList.map((e) => e.map((f) => Field(f)))
+        );
+        latestActionsHash = SequenceEvents.updateSequenceState(
+          latestActionsHash,
+          actionsHash
+        );
+        actionsList.push({
+          actions: currentActionList,
+          hash: Ledger.fieldToBase58(Field(latestActionsHash)),
         });
+        currentAccountUpdateId = accountUpdateId;
+        currentActionList = [data];
       } else {
-        actionMap.set(accountUpdateId, {
-          hash: Ledger.fieldToBase58(Field(actionState)),
-          actions: [data],
-        });
+        currentActionList.push(data);
       }
-    }
-  }
+    });
+
+    const currentActionHash = Ledger.fieldToBase58(Field(latestActionsHash));
+    const expectedActionHash = Ledger.fieldToBase58(Field(actionState));
 
-  // Reverse the order of actions since the API returns in descending order of block height while Localblockchain pushes new actions to end of array.
-  const actionData = [...actionMap.values()].reverse();
+    if (currentActionHash !== expectedActionHash) {
+      throw new Error(
+        `Failed to derive correct actions hash for ${publicKey}. Derived hash: ${currentActionHash}, expected hash: ${expectedActionHash})`
```

Comment:
I wonder if we can tell the user what to do in that case: like "try a different archive API" or maybe the previous tx hasnt been included yet?

Reply:
Maybe also let the user know what actions we found so its easier to debug

Reply:
Will do!